File: python/1006_笨阶乘_medium.py
class Solution:
    def clumsy(self, N: int) -> int:
        if N <= 1:
            return 1 if N == 1 else 0
        operator = {
            0: mul,
            1: lambda x, y: int(x/y),
            2: add,
            3: sub
        }
        ans = N
        ansList = []
        for i in range(N, 1, -1):
            if (N-i) % 4 == 3:
                if not ansList:
                    ansList.append(ans)
                else:
                    ansList.append(-ans)
                ans = i-1
            elif (N-i) % 4 == 2:
                if not ansList:
                    ans += i-1
                else:
                    ansList[0] += i-1
            else:
                ans = operator[(N-i)%4](ans, i-1)
        if ansList:
            return sum(ansList)-ans
        else:
            return ans

File: cpp/1006_笨阶乘_medium.cpp
class Solution {
public:
    int clumsy(int N) {
        stack<int> stk;
        stk.push(N);
        N--;
        int index = 0;
        while (N) {
            if (index%4 == 0)
                stk.top() *= N;
            else if (index%4 == 1)
                stk.top() /= N;
            else if (index%4 == 2)
                stk.push(N);
            else
                stk.push(-N);
            index++;
            N--;
        }
        int ans = 0;
        while(!stk.empty()){
            ans += stk.top();
            stk.pop();
        }
        return ans;

    }
};